src/main/java/io/pinecone/clients/Pinecone.java

```diff
@@ -867,7 +873,73 @@ public Builder withSourceTag(String sourceTag) {
          * @return This {@link Builder} instance for chaining method calls.
          */
         public Builder withOkHttpClient(OkHttpClient okHttpClient) {
-            this.okHttpClient = okHttpClient;
+            this.customOkHttpClient = okHttpClient;
+            isCustomOkHttpClient = true;
+            return this;
+        }
+
+        /**
+         * Sets a proxy for the Pinecone client to use for control plane requests.
+         * <p>
+         * When a proxy is configured using this method, all control plane requests made by the Pinecone client will be routed
+         * through the specified proxy server.
+         * <p>
+         * It's important to note that both proxyHost and proxyPort parameters should be provided to establish
+         * the connection to the proxy server.
+         * <p>
+         * Example usage:
+         * <pre>{@code
+         *
+         * String proxyHost = System.getenv("PROXY_HOST");
+         * int proxyPort = Integer.parseInt(System.getenv("PROXY_PORT"));
+         * Pinecone pinecone = new Pinecone.Builder("PINECONE_API_KEY")
+         *     .withControlPlaneProxy(proxyHost, proxyPort)
+         *     .build();
+         *
+         * // Network requests for control plane operations will now be made using the specified proxy.
+         * pinecone.listIndexes();
+         * }</pre>
+         *
+         * @param proxyHost The hostname or IP address of the proxy server. Must not be null.
+         * @param proxyPort The port number of the proxy server. Must not be null.
+         * @return This {@link Builder} instance for chaining method calls.
+         */
+        public Builder withControlPlaneProxy(String proxyHost, int proxyPort) {
+            this.controlPlaneProxyConfig = new ProxyConfig(proxyHost, proxyPort);
+            Proxy proxy = new Proxy(Proxy.Type.HTTP, new InetSocketAddress(proxyHost, proxyPort));
+            okHttpClientBuilder.proxy(proxy);
+            return this;
+        }
```

Comment:
I'm wondering how this approach of adding named proxy-overrides (e.g. "control plane proxy", "data plane proxy", etc.) scales as we continue to add more API surface e.g. Inference API is coming soon, and probably Knowledge Models too when it makes sense. 

Instead, I'm wondering if it makes sense to keep it simple:

```java
        Pinecone pinecone = new Pinecone.Builder(apiKey)
                .withProxy(proxyHost, proxyPort)
                .build();

        System.out.println(pinecone.describeIndex("PINECONE_INDEX"));
```

and if some really needs to set a different proxy for control plane vs data plane then they would do it by instantiating a different client instance:

```java
        Pinecone controlPinecone = new Pinecone.Builder(apiKey)
                .withProxy(controlProxyHost, controlProxyPort)
                .build();

        Pinecone dataPinecone = new Pinecone.Builder(apiKey)
                .withProxy(dataProxyHost, dataProxyPort)
                .build();
```

Reply:
That's a great point! I started off with .withProxy(String host, String port) as well but given the control and data plane rely on OkHttpClient and NettyChannelBuilder respectively, we must pass in two extra flags to identify which operation needs the proxy to be configured with. We could have used enum but I was under the impression that enum can only allow us to choose one value at any time (which is still correct) but I just learned about varargs so my proposed design is now:


```java
public Builder withProxy(String proxyHost, int proxyPort, ProxyType... proxyTypes) {
    for (ProxyType proxyType : proxyTypes) {
        switch (proxyType) {
            case CONTROL_PLANE:
                this.controlPlaneProxyConfig = new ProxyConfig(proxyHost, proxyPort);
                Proxy proxy = new Proxy(Proxy.Type.HTTP, new InetSocketAddress(proxyHost, proxyPort));
                okHttpClientBuilder.proxy(proxy);
                break;
            case DATA_PLANE:
                this.dataPlaneProxyConfig = new ProxyConfig(proxyHost, proxyPort);
                break;
            case INFERENCE_API:
                this.inferenceApiProxyConfig = new ProxyConfig(proxyHost, proxyPort);
                // Additional configuration for inference API if necessary
                break;
            case KNOWLEDGE_MODEL:
                this.knowledgeModelProxyConfig = new ProxyConfig(proxyHost, proxyPort);
                // Additional configuration for knowledge model if necessary
                break;
        }
    }
    return this;
}
```


So the usage will look like this:

```java
// Configure control plane only
Pinecone pinecone = new Pinecone.Builder(apiKey)
        .withProxy(proxyHost, proxyPort, ProxyType.CONTROL_PLANE)
        .build();

// Configure data plane only
Pinecone pinecone = new Pinecone.Builder(apiKey)
        .withProxy(proxyHost, proxyPort, ProxyType.DATA_PLANE)
        .build();

// Configure both control and data plane
Pinecone pinecone = new Pinecone.Builder(apiKey)
        .withProxy(proxyHost, proxyPort, ProxyType.CONTROL_PLANE, ProxyType.DATA_PLANE)
        .build();

// Configure neither control or data plane
Pinecone pinecone = new Pinecone.Builder(apiKey)
        .withProxy(proxyHost, proxyPort)
        .build();
```

Although I'm not yet convinced with the enum name ProxyType, given it can mean actual proxy types. Thoughts?

Reply:
That's a great point! I started off with .withProxy(String host, String port) as well but given the control and data plane rely on OkHttpClient and NettyChannelBuilder respectively, we must pass in two extra flags to identify which operation needs the proxy to be configured with. We could have used enum but I was under the impression that enum can only allow us to choose one value at any time (which is still correct) but I just learned about varargs so my proposed design is now (added placeholder for Inference api and knowledge model in the sample below but won't be in the codebase):


```java
public Builder withProxy(String proxyHost, int proxyPort, OperationType... operationTypes) {
    for (OperationType operationType : operationTypes) {
        switch (operationType) {
            case CONTROL_PLANE:
                this.controlPlaneProxyConfig = new ProxyConfig(proxyHost, proxyPort);
                Proxy proxy = new Proxy(Proxy.Type.HTTP, new InetSocketAddress(proxyHost, proxyPort));
                okHttpClientBuilder.proxy(proxy);
                break;
            case DATA_PLANE:
                this.dataPlaneProxyConfig = new ProxyConfig(proxyHost, proxyPort);
                break;
            case INFERENCE_API:
                this.inferenceApiProxyConfig = new ProxyConfig(proxyHost, proxyPort);
                break;
            case KNOWLEDGE_MODEL:
                this.knowledgeModelProxyConfig = new ProxyConfig(proxyHost, proxyPort);
                break;
        }
    }
    return this;
}
```


So the usage will look like this:

```java
// Configure control plane only
Pinecone pinecone = new Pinecone.Builder(apiKey)
        .withProxy(proxyHost, proxyPort, OperationType.CONTROL_PLANE)
        .build();

// Configure data plane only
Pinecone pinecone = new Pinecone.Builder(apiKey)
        .withProxy(proxyHost, proxyPort, OperationType.DATA_PLANE)
        .build();

// Configure both control and data plane
Pinecone pinecone = new Pinecone.Builder(apiKey)
        .withProxy(proxyHost, proxyPort, OperationType.CONTROL_PLANE, OperationType.DATA_PLANE)
        .build();

// Configure neither control or data plane
Pinecone pinecone = new Pinecone.Builder(apiKey)
        .withProxy(proxyHost, proxyPort)
        .build();
```

```java
public enum OperationType {
    CONTROL_PLANE,
    DATA_PLANE
    // KNOWLEDGE_MODEL
    // INFERENCE_API
}
```

Reply:
Could we just optimistically configure both Ok client and Netty Bldr with the proxy config? 

I still feel like the different (operation types) X (proxy configs) could be hard to reason about especially as the API surface areas grow.

Reply:
After discussing internally, we have decided to go with @ssmith-pc's approach of configuring both OkHttpClient and NettyChannelBuilder if .withProxy() is called instead of providing .withControlPlaneProxy and .withDataPlaneProxy since end user can instantiate Pinecone client twice with .withProxy(), one for control plane proxy url and the other for data plane proxy url if it comes to the point where the user needs to have two separate proxies for control and data plane operations.